Add French Email Translations

momo3404 opened the pull request:
Fixes French emails not being sent to French users.

Changes proposed in this PR:
- Add French translations to email views.
- Add french email signature.

momo3404 committed:
Added French emails

momo3404 committed:
Fixed brakeman errors

aaronskiba commented on app/views/user_mailer/_french_email_signature.html.erb:
```diff
@@ -0,0 +1,24 @@
+<p>
+  <%= 'Sincèrement,' %>
+  <br />
+  <%= 'Équipe chargée de l’outil %{tool_name}' %{ tool_name: tool_name} %>
```
We should translate the `tool_name` ('DMP Assistant') references to French.

momo3404 replied:
Would just hardcoding `Assistant PGD` in the French ones be acceptable?

aaronskiba reviewed (commented):
If we're using this approach, I guess we'll also now have to display all of the `subject: ` text in both languages (`app/mailers/user_mailer.rb`).

aaronskiba commented:
`I18n.with_locale I18n.locale do` is used throughout `app/mailers/user_mailer.rb` and the English views have `_()` enabled. So I'm thinking what will happen if French is selected within the app and then an email is delivered. Rather than English / French, the email would probably appear as French / French.

momo3404 committed:
Added subject translations

momo3404 commented:
> `I18n.with_locale I18n.locale do` is used throughout `app/mailers/user_mailer.rb` and the English views have `_()` enabled. So I'm thinking what will happen if French is selected within the app and then an email is delivered. Rather than English / French, the email would probably appear as French / French.

Would the easiest fix for this be to just remove the '_()' from the English email text?

momo3404 committed:
Added translations in controller, removed unnecessary email translations

aaronskiba commented on app/mailers/user_mailer.rb:
```diff
@@ -22,7 +22,7 @@ def welcome_notification(user)
     I18n.with_locale I18n.locale do
       mail(to: @user.email,
-           subject: format(_('Welcome to %{tool_name}'), tool_name: tool_name))
+           subject: format("Welcome to %{tool_name} / Bienvenue sur l'%{tool_name}", tool_name: tool_name))
```
Because we want both "DMP Assistant" and "Assistant PGD" here, maybe we'll need to hardcode? The same is true for the sharing notification.

aaronskiba replied:
I wonder if we should just remove the `I18n.with_locale I18n.locale do` from both of those places too.

momo3404 replied:
Yea I think hardcoding it in those two messages makes sense, and what would we replace the `I18N' calls in those lines with?

aaronskiba replied:
I'm thinking the following might be a clean approach:
Remove the  `I18n.with_locale I18n.locale do`  and `end` code that surrounds the block for `app/mailers/user_mailer.rb`.

Within `app/views/user_mailer/welcome_notification.html.erb` (and app/views/devise/mailer/invitation_instructions.html.erb), surround the existing code with the following:

```ruby
<% I18n.available_locales.each do |locale| %>
  <% I18n.with_locale(locale) do %>
  ...CODE PRIOR TO THIS PR...
  <% end %>
<% end %>
```

aaronskiba replied:
I could add a PR that attempts that if you'd like.

momo3404 replied:
That looks cleaner, go for it!

aaronskiba commented on app/mailers/user_mailer.rb:
```diff
@@ -211,7 +214,7 @@ def api_credentials(api_client)
     @helpdesk_email = helpdesk_email(org: @api_client.org)
 
-    I18n.with_locale I18n.locale do
+    I18n.with_locale(@api_client.language&.abbreviation || I18n.default_locale) do
```
I guess there's no language column for the `api_clients` table. Not totally sure if this one will require the same English / French approach. Or maybe we could use something like `@api_client.org.language.abbreviation`.

momo3404 replied:
Im thinking some users in an org might have a different preferred language. So I think the english/french email will be better here. Unfortunately we'll have to wait for translations for this one.

aaronskiba replied:
Good point, the recipient's preferred language might differ from the org's.

aaronskiba commented on app/views/user_mailer/sharing_notification.html.erb:
```diff
@@ -1,5 +1,5 @@
 <p>
```
I was testing the email relay to trigger this view by sending out an email on the `/plans/:id/share` path. However, nothing seemed to happen.

It looks like the view we want to modify is actually at `app/views/devise/mailer/invitation_instructions.html.erb`. For some extra context, the view must be getting triggered at the following:
```ruby
# app/controllers/roles_controller.rb
            User.invite!({ email: role_params[:user][:email],
                           firstname: _('First Name'),
                           surname: _('Surname'),
                           org: current_user.org },
                         current_user)
```

aaronskiba replied:
It looks like the corresponding email title needs to be updated here:

```ruby
# app/models/user.rb

  # Override devise_invitable email title
  def deliver_invitation(options = {})
    super(options.merge(subject: format(_('A Data Management Plan in %{application_name} has been shared with you'),
                                        application_name: ApplicationService.application_name))
    )
  end
```

aaronskiba replied:
After debugging a little more, we figured out the following:
- `app/views/devise/mailer/invitation_instructions.html.erb` (and `User.invite!` within `app/controllers/roles_controller.rb`) is rendered when the email being invited does not yet exist in the db.
- `app/views/user_mailer/sharing_notification.html.erb` is rendered when the email does exist in the db.

aaronskiba replied:
Given this, we can probably revert the changes to `app/views/user_mailer/sharing_notification.html.erb`, and just update `I18n.with_locale I18n.locale do` in `def sharing_notification` to align with the invitee's language.

momo3404 committed:
Correct sharing email

momo3404 committed:
Added French app name

momo3404 committed:
Added changelog entry, made rubocop happy

aaronskiba committed:
Refactor email bodies that render both En & Fr

`app/views/devise/mailer/confirmation_instructions.html.erb`
- The confirmation instructions are delivered at the time of account creation. Because a user has not yet had the option to select their preferred language, we are now rendering this email body in both English and French.
- NOTE: Because we have recently re-implemented the `:confirmable` module, many existing users also have to confirm their emails. If we wanted, we could use `user.language` for determining the language of the confirmation instructions for these users.

`app/views/devise/mailer/invitation_instructions.html.erb` & `app/views/user_mailer/welcome_notification.html.erb`
- Both of these emails are sent to users who have not yet had a chance to select a preferred language within the app. As a result, we are rendering these emails in both English and French.
- This change refactors the English and French rendering by iterating through `I18n.available_locales` and translating the base message to each available language.

aaronskiba committed:
Render email confirmation subject to be EN / FR

This change updates the `confirmation_instructions` email title to be in EN / FR format.

aaronskiba committed:
Delete _french_email_signature.html.erb

aaronskiba committed:
Fix/enable translations in EN & FR emails

- These changes enable the translation of `tool_name` (see app/helpers/mailer_helper.rb#tool_name)
- For `app/views/devise/mailer/invitation_instructions.html.erb`, needed translations are also moved inside the `I18n.available_locales` loop.

aaronskiba committed:
Fix/enable translations in reset password email

- Updates `I18n.with_locale I18n.locale` to `I18n.with_locale(user.language&.abbreviation || I18n.default_locale)`
- Moves translating of `tool_name` and `email_subject` inside the `I18n.with_locale I18n.locale` to `I18n.with_locale(user.language&.abbreviation || I18n.default_locale)` block.

aaronskiba committed:
Remove locale set  in welcome_notification action

Given the EN & FR locale handling within `app/views/user_mailer/welcome_notification.html.erb` (see 399f5f628), the locale setting can be removed here.

momo3404 committed:
Merge pull request #1096 from portagenetwork/momo/email-translation-copy

Refactor Email Bodies that Render Both English & French

momo3404 committed:
Fix feedback emails not sending

aaronskiba committed:
Hardcode `welcome_notification` email subject

We need `tool_name` to be both in English and French for this subject line. Hardcoding is a simple way to achieve that.

aaronskiba committed:
Update user_mailer_spec.rb after 5d14a8ff1

- This update now aligns with 5d14a8ff1, which hardcodes the subject line of the delivered `UserMailer.welcome_notification`.

aaronskiba committed:
Fix welcome_notification @email_subject translation

`app/views/user_mailer/welcome_notification.html.erb` renders the email body both in English and French. This change ensures that `@email_subject` is properly translated for both languages.

momo3404 committed:
Change default_locale to locale

momo3404 committed:
Merge branch 'momo/email-translation' of https://github.com/portagenetwork/roadmap into momo/email-translation

aaronskiba committed:
Match privileges_list language to email recipient

This change moves `privileges_list(@user)` inside of the `I18n.with_locale(@user.language&.abbreviation || I18n.locale)` block, ensuring the correct language is rendered.

aaronskiba committed:
Refactor setting of email locale

aaronskiba committed:
Merge pull request #1099 from portagenetwork/email-translation-refactor

Refactor setting of email locale
